### ISSUE TITLE
batch-represent creates duplicate labels

### ISSUE THREAD

**Issue by shimen** [cmusatyalab/openface#160]
### Context of the issue.
When testing with batch-represent option on my data I got missing representations.
Duplicate labels were created. When I had checked the labels file for duplicates I got 175 duplicate paths from 4942 files.

I had printed all the paths to files.
In batch-represent.lua I had added these commands:

   for i=1,nImgs do
	 print ( ffi.string(dumpLoader.imagePath[dumpLoader.testIndices[i]]:data()))
   end

   for i=1,nImgs do
	 print ( ffi.string(dumpLoader.imagePath[i]:data()))
   end

The first for loop is the current version for getting the file path. It depends on the dumpLoader.testIndices[i].
here is the output file:
[labels_duplicates.txt](https://github.com/cmusatyalab/openface/files/359325/labels_duplicates.txt)

The second for loops over all the indices. no duplicates are found
here is the output file:
[labels_all files.txt](https://github.com/cmusatyalab/openface/files/359326/labels_all.files.txt)

Have you encountered this problem?

### Steps to reproduce.
It is usually does not happens, most of the times just on large folders of pictures.
I'm not sure how to reproduce it.

### OS and hardware information.

+ Operating system: centos 6.7
+ Torch version: 7
+ CPU architecture: [todo]
+ GPU type (if using):  no gpu

-Ilya

**Comment by bamos** [comment]
Hi Ilya, interesting, thanks for the report. I haven't noticed this before. I used `dataset.lua` in `batch-represent` out of convenience. I think it's better to replace it since `batch-represent` doesn't need most of the features and just needs to iterate through a dataset. Potentially https://github.com/torchnet/torchnet could be useful. Let me know if you're interested in helping.

The training code uses a similar `dataset.lua`, but from my understanding this shouldn't hurt the convergence (other than leaving some images out).

-Brandon.

**Comment by shimen** [comment]
Hi Brandon,
I just managed to find the problem.
Can you please test a small program on your environment. I had attached a 'test.txt' file change it to 'test.lua'.
[test.txt](https://github.com/cmusatyalab/openface/files/359670/test.txt)

Here is what I found out. The problem for me is with the combination of "torch.setdefaulttensortype('torch.FloatTensor')" with the  linspace command for certain values (4942) for 4943 it works fine.

This is the output with enabled "torch.setdefaulttensortype('torch.FloatTensor')"
  4759 : 4934
  4760 : 4935
  4761 : 4936
  4762 : 4937
  4763 : 4938
  4764 : 4939
  4765 : 4940
  4766 : 4941
  4767 : 4942

4767 unique values instead of 4942

This is the output with disabled "torch.setdefaulttensortype('torch.FloatTensor')"
  4934 : 4934
  4935 : 4935
  4936 : 4936
  4937 : 4937
  4938 : 4938
  4939 : 4939
  4940 : 4940
  4941 : 4941
  4942 : 4942

now it is correct!

Can you check that you have the same problem?

torch.linspace is used in the 'dataset.lua' file


-Ilya

**Comment by bamos** [comment]
Hi Ilya,

Thanks, very interesting! I see the same behavior in the file you sent. It works with double tensors (the default), but not with floats. Replacing `torch.linspace` with `torch.range(1, endIndex):long()` fixes this on my machine, I think due to floating point rounding errors in `torch.linspace` for large ranges. I have updated `dataset.lua` to use `torch.range`. Can you pull the latest code and try on your dataset again?

-Brandon.

**Comment by shimen** [comment]
Hi Brandon, now it is working as expected!

Thank you,
-Ilya
